PayloadPkg/OsLoader/OsLoader.c

```diff
@@ -154,8 +143,126 @@ ParseIasImage (
     return EFI_UNSUPPORTED;
   }
 
+  return Status;
+}
+
+/**
+  Parse PODS image
+
+  This function will parse PODS image and get every file blobs info, Especially
+  get the command line file and ELF/kernel file for boot.
+
+  @param[in]      BootOption    Current boot option
+  @param[in, out] LoadedImage   Loaded Image information.
+
+  @retval  RETURN_SUCCESS       Parse IAS image successfully
+  @retval  Others               There is error when parsing IAS image.
+**/
+EFI_STATUS
+ParseContainerImage (
+  IN     OS_BOOT_OPTION      *BootOption,
+  IN OUT LOADED_IMAGE        *LoadedImage
+  )
+{
+  EFI_STATUS                  Status;
+  CONTAINER_HDR              *ContainerHdr;
+  UINT64                      ComponentName;
+  LOADER_COMPRESSED_HEADER   *LzHdr;
+  IMAGE_DATA                  File[MAX_IAS_SUB_IMAGE];
+  UINT8                       Index;
+
+  ContainerHdr = (CONTAINER_HDR  *)LoadedImage->IasImage.Addr;
+  Status = RegisterContainer ((UINT32) ContainerHdr);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "Image given is not a valid PODS image\n"));
+    return EFI_LOAD_ERROR;
+  }
+
+  AddMeasurePoint (0x4080);
+  ZeroMem (File, sizeof (File));
+
+  DEBUG ((DEBUG_INFO, "PODS size = 0x%x, image type = 0x%x, # of components = %d\n", LoadedImage->IasImage.Size, ContainerHdr->ImageType, ContainerHdr->Count));
+
+  // Enumerate all components
+  Index = 0;
+  ComponentName = 0;
+  while (TRUE) {
+    Status = GetNextAvailableComponent (ContainerHdr->Signature, (UINT32 *)&ComponentName);
+    if (EFI_ERROR(Status) || ((UINT32)ComponentName == SIGNATURE_32 ('_', 'S', 'G', '_')) ) {
+      break;
+    } else {
+      DEBUG ((DEBUG_INFO, "COMP:%a %r\n", &ComponentName, Status));
+      // Load or locate components
+      if (ContainerHdr->Flags & CONTAINER_HDR_FLAG_MONO_SIGNING) {
+        // Use Locate to reuse the image buffer
+        Status  = LocateComponent (ContainerHdr->Signature, (UINT32) ComponentName, (VOID **)&LzHdr, NULL);
+        if (!EFI_ERROR(Status)) {
+          if (LzHdr->Signature != LZDM_SIGNATURE) {
+            Status = EFI_UNSUPPORTED;
+          } else {
+            File[Index].Addr = (UINT8 *)LzHdr + sizeof(LOADER_COMPRESSED_HEADER);
+            File[Index].Size = LzHdr->Size;
+          }
+        }
+      } else {
+        // Use Load to decompress to a new buffer
+        Status = LoadComponent (ContainerHdr->Signature, (UINT32) ComponentName, (VOID **)&File[Index].Addr, &File[Index].Size);
+      }
+    }
+    Index++;
```

Comment:
Pls consider error condition.  Index should only be increased Status is SUCCESS.